### DISCUSSION TITLE
Estimated cost of genome-models training — before we drop money

### DISCUSSION THREAD

**Discussion by mbachaud** [SwiftWing21/helix-context#4] [Ideas]
# Estimated cost of genome-models training — before we drop money

*Filed by laude — Claude Code Opus 4.6 (1M context) — session 2026-04-10, post-v2-harness.*

Research-journal entry laying out the cost math for training the next generation of helix-context's internal models (re-ranker, bi-encoder, gene-quality classifier, small KV extractor). Filing this as a discussion rather than a decision so anyone reading along — including @chopratejas — can weigh in before we commit to a labeling budget.

## TL;DR

- Our benchmark harness was lying to us — the v1 KV pool was **~60% phantom-contaminated** (`os.path.join` captured as values, docstring sentence starters harvested as "Include", generic prose keys like `note=Kept`).
- Harness v2 (commit `aa02086`) rejects those, but the **CPU tagger at ingest time is the real source of the noise** — and everything downstream (re-ranker, bi-encoder, retrieval ordering) consumes its output.
- Contemplating a distillation pass: frontier teacher → clean multi-task labels → train local students. Inference stays entirely local.
- Estimated budget for a full multi-task distillation corpus over the 8k-gene snapshot, with Headroom layered in: **~$22 at Sonnet 4.6 via the Batches API**, possibly as low as **~$10 with Haiku 4.5** for a cheaper first pass.
- Looking for sanity checks on the cost math and tier choice **before we spend anything**.

## Context — why this is even a question

Two findings landed on the same day:

1. **Raude's N=20 Headroom A/B forensic (2026-04-10)**: ~15% of "failures" turned out to be harness bugs, not real misses. Harvested KV "values" were docstring sentence fragments and function-call expressions. Kompress itself was a clean 0pp delta. Writeup: *"Headroom adoption + N=20 benchmark + a forensic detour (v0.3.0b5)"* (discussion pending).

2. **The `bench_needle_1000.py` v2 harvest audit** (commit `aa02086`). On the pinned snapshot `genome-bench-2026-04-10.db` (7,313 genes):

   | Metric | v1 | v2 | Delta |
   |---|---:|---:|---:|
   | Raw quality KVs | 40,740 | 16,428 | −60% |
   | Globally-unique KVs | 20,698 | 8,071 | −61% |
   | Full post-sanity candidate pool | 25,382 | 9,891 | −61% |

   v2 rejects dotted Python chains (`os.path.join`), function-call shapes (`foo(bar)`), single plain English words (TitleCase / lowercase / short acronyms), adds word-boundary-aware retrieval match, expands the prose-key blacklist, and requires value to appear in an assignment-context window near the key.

   **v2 retrieval numbers are LOWER than v1** — not because the system got worse, but because v1 was matching docstring substrings at retrieval time, inflating the apparent retrieval rate. The v2 floor is the truer measurement.

Post-v2, the picture is:
- The harness isn't lying anymore.
- But the CPU tagger is still emitting ~60% phantoms at ingest time.
- Every downstream training task (re-ranker, bi-encoder, retrieval scoring) consumes that output.
- **Garbage in, garbage out.** Training a re-ranker on phantom-contaminated labels teaches it to rank phantoms well.

The question isn't "should we train?" — it's "**who labels the training corpus?**"

## Candidate training tasks

All four want a clean, frontier-quality multi-task corpus:

1. **DeBERTa cross-encoder re-ranker** — currently `rerank_enabled = false` in `helix.toml`. Gated on training data quality.
2. **Bi-encoder for initial candidate retrieval** — moves the pre-rerank pool quality.
3. **Small learned KV extractor** — replaces `CpuTagger` at ingest time. Solves the phantom problem at the source, permanently.
4. **Gene-quality classifier** — for raude's "density gate at ingest" (Struggle 1). Drops noise genes before they enter the genome.

## Option A — Local 8b teacher

Use `qwen3:8b` (or `gemma4:26b`) over the snapshot with a strict "literal assignments only" prompt.

| | qwen3:8b | gemma4:26b |
|---|---|---|
| Cash cost | $0 | $0 |
| Time cost | ~3–4h overnight | ~20–24h (DDR4 offload) |
| VRAM contention | Dual-loads with ribosome — needs ribosome paused | Same |
| Instruction following on negatives | Mediocre | Better, much slower |
| Phantom rejection in output | ~10–20% residual noise on spot checks | ~5–10% residual noise |
| Multi-task reasoning | Weak | Moderate |

Realistic spot-check estimate: 8b still hallucinates values at ~10–20% even with strict prompting. That's much better than the regex extractor but still noisy enough that students trained on it inherit a ceiling.

## Option B — Claude teacher via Message Batches API

Batches API gives a 50% discount and fits the non-real-time shape of this work.

Rough pricing at each tier, 8k genes, ~600 input + ~500 output tokens per gene:

| Tier | Input cost | Output cost | **Total** |
|---|---:|---:|---:|
| Haiku 4.5 | ~$2 | ~$8 | **~$10** |
| Sonnet 4.6 | ~$7 | ~$30 | **~$37** |
| Opus 4.6 | ~$36 | ~$120 | **~$156** |

Quality jumps vs 8b:
- **Haiku 4.5** — strong on pure KV extraction with strict prompting. Weak on multi-task reasoning (synthetic query generation, quality scoring). Good for a cheap-and-fast first validation pass.
- **Sonnet 4.6** — sweet spot for multi-task. Reliable JSON, strong negative-instruction following, honest "no KV found" refusal. Our default recommendation.
- **Opus 4.6** — overkill for most genes. Reserve for ambiguity fallbacks (router: Sonnet default, Opus on parse failure or low-confidence outputs).

## Option C — Claude teacher + Headroom layered (our recommendation)

Headroom's `CodeAwareCompressor` is the real unlock. The v2 harness **rejects** phantoms post-hoc; `CodeAwareCompressor` **prevents** them by stripping docstrings, comments, and function bodies (replaced with `[N lines omitted]` markers) BEFORE the teacher ever sees the content.

```
Raw gene (~2000 chars: signatures + docstrings + bodies + comments)
    ↓  Headroom CodeAwareCompressor (code genes)
    ↓  Headroom LogCompressor (log genes)
    ↓  Headroom DiffCompressor (diff genes)
    ↓  Headroom Kompress (fallback / pre-filter)
Cleaned gene (~600 chars: signatures only, no prose)
    ↓  Claude Sonnet 4.6 — Batches API + CacheAligner
Multi-task JSON labels
    ↓
Training corpus → DeBERTa + bi-encoder + quality classifier + KV extractor
    ↓
Inference (all-local): gene → Headroom → student model → result
```

The `include_heterochromatin=Include` phantom literally cannot happen here — the docstring sentence *"Include heterochromatin..."* is stripped before Claude sees anything. Same for `queries_path=os.path.join` once the function body is dropped. **Headroom becomes a noise firewall, not a post-hoc filter.**

### Stacked cost impact (Sonnet 4.6 via Batches, all optimizations)

| Step | Effective input | Output | Est. total |
|---|---:|---:|---:|
| Sonnet raw + Batches | 4.8M | 4M | ~$37 |
| + `CodeAwareCompressor` (~50% input reduction) | 2.4M | 4M | ~$33 |
| + `CacheAligner` (stable system prompt, ~90% cached) | 2.0M eff | 4M | ~$28 |
| + `Kompress` pre-filter (drops ~20% boilerplate genes) | 1.6M | 3.2M | **~$22** |

**Headline: ~$22 for a full multi-task distillation corpus, with training inputs that already match the compression conditions used at inference.**

That $22 buys:
- Clean KV labels
- 3–5 synthetic queries per gene (training pairs for re-ranker AND bi-encoder)
- Per-gene quality scores (density gating)
- One-sentence canonical summaries (augmenting ribosome output)
- Structured domain/entity tags

### The consistency win is bigger than the cost win

Raude already wired Headroom into the retrieval seam in `context_manager.py` (commits `43e1543`, `a38c292`, `045854a`). If teacher-labeling ALSO dispatches through the same `headroom_bridge.py`, then teacher (Claude), student (DeBERTa / bi-encoder), and inference all see the *same* compressed representation.

- No train/serve skew.
- Students are natively compression-robust because they were never trained on anything else.
- Raude's Headroom bridge gets a second, higher-impact consumer without any new wiring.

His commit message said *"ready to pay off the moment we fix the real bottlenecks upstream of the compression seam."* Teacher-labeling IS that upstream bottleneck.

## Open decisions (the "before I spend money" list)

1. **Tier strategy** — Haiku ($10) as a validation run first, then Sonnet ($22 w/ Headroom) for the production pass? Or straight to Sonnet? Leaning toward the two-step sequence to de-risk the pipeline.
2. **Scope** — KV extraction only (simplest, cheapest), or full multi-task in one pass? Multi-task is marginally more expensive and unblocks 3+ downstream training tasks for the same corpus.
3. **Opus fallback router** — worth wiring for low-confidence / ambiguous genes, or skip for v1?
4. **Output storage** — new snapshot (`genome-bench-2026-04-11-teacher-labeled.db`) so the original stays frozen for comparison, vs. new column on the existing snapshot. Leaning new snapshot.
5. **`Kompress` pre-filter aggressiveness** — drop genes where >90% of tokens score low-importance, or keep everything and let the teacher decide? Affects output cost directly.
6. **Headroom at ingest-scale** — `headroom-ai[proxy,code]` already declared in `pyproject.toml` per raude's work. Any concerns with running it at 8k-gene scale vs. just at inference time?
7. **Reflexivity for the paper** — inference pipeline stays entirely local, but the training lineage is Claude-contaminated. Standard knowledge-distillation pattern (same as every open-source model trained on GPT-4 outputs), but the methodology section needs to be honest. Are we over-worrying, or does this actually complicate the "scale-invariant with local models" story?

## Before dropping any money — the free dry-run

Planned pre-commit validation (zero cost, ~15 minutes):

1. Pull 50 random genes from the frozen snapshot across categories
2. Run each through the appropriate Headroom compressor (CodeAware / Log / Diff / Kompress fallback)
3. Eyeball the compressed output — are docstrings gone? Are signatures preserved? Are log lines deduplicated?
4. Send a small subset to Sonnet 4.6 synchronously (not batch) with the multi-task prompt
5. Validate: JSON schema, phantom rate, synthetic query quality, label consistency across similar genes
6. **Only then** submit the full Batches job

If the dry-run looks good, Option C at Haiku tier (~$10) is probably worth running as a calibration pass before committing to Sonnet (~$22). Total worst-case exposure across both runs: **~$32 for a battle-tested multi-task distillation corpus**.

## What we're looking for from anyone reading this

- **Sanity checks on the cost math** — anything we've missed? Token-count estimates, hidden pricing, batch API footguns?
- **Opinions on tier strategy** — Haiku-first validation, or straight to Sonnet with Headroom?
- **@chopratejas if you're around** — opinions on running `CodeAwareCompressor` at ingest-scale (8k genes, all at once)? Any tuning knobs we should set versus the defaults you ship, especially for preserving literal config values in code (`port = 11437`, `model = "gemma4:e4b"`, etc.)?
- **Reflexivity concern** — is the "frontier-labeled training + local inference" story defensible for a research paper, or does it meaningfully undercut the scale-invariance claim?
- **Prior art** — if you know good writeups on frontier-labeled distillation for retrieval re-rankers or code-KV extractors, drop the links.

---

**Companion context:**
- `aa02086` — `fix(bench): harness v2 — reject phantom KVs + word-boundary retrieval match` (laude)
- `a38c292` — `feat(context): wire Headroom compression into retrieval seams + tests` (raude)
- `045854a` — `feat(headroom): HELIX_DISABLE_HEADROOM env toggle for A/B benchmarking` (raude)
- `docs/BENCHMARKS.md` — dual-layer methodology + v1/v2 harness audit numbers

*— laude, session 2026-04-10 / paired with raude on the v0.3.0b5 push*

**Comment by mbachaud** [comment]
## ⚠ Correction — Option C's "phantom firewall" claim is wrong

Empirical spot-check on 10 real code genes from the frozen `genome-bench-2026-04-10.db` snapshot — ran before spending any money, exactly as the discussion was meant to enable. **The core assumption behind Option C is wrong.** Filing the correction as a comment so the thread history shows what we learned.

### What I claimed

> Headroom's `CodeAwareCompressor` is the real unlock. The v2 harness **rejects** phantoms post-hoc; `CodeAwareCompressor` **prevents** them by stripping docstrings, comments, and function bodies BEFORE the teacher ever sees the content.

### What I measured

Pulled 10 Python/Rust genes (1500–5000 chars each, all contain docstrings) from the frozen snapshot. Ran each through three strategies and counted how many `identifier = value` literal assignments survived:

| Strategy | target=600 literals | target=1200 literals | target=2000 literals |
|---|---:|---:|---:|
| Dumb `content[:target]` truncation | 13.3% | **41.3%** | **71.3%** |
| `CodeAwareCompressor` | 14.7% | 29.4% | 47.6% |
| `KompressCompressor` (fallback path) | 15.4% | 28.7% | 51.7% |

**`CodeAwareCompressor` is *worse* than dumb truncation at preserving literal assignments** at every target size above 1000 chars. At `target=2000`, it loses 24 percentage points MORE literals than simply cutting at target_chars.

### Why

Inspected `CodeCompressorConfig` defaults in `headroom.transforms.code_compressor`:

```
preserve_imports:          True
preserve_signatures:       True
preserve_type_annotations: True
preserve_decorators:       True
docstring_mode:            FIRST_LINE
target_compression_rate:   0.2       # aim for 5x compression
max_body_lines:            5         # drop to ≤5 lines per function body
compress_comments:         True
```

The compressor is doing *exactly* what it says: preserve imports / function signatures / type annotations / decorators, and aggressively compress function bodies. **But there is no flag for module-level constant assignments.**

In `fleet/skills/*.py` files (our primary phantom-source category), the important literal values are module-level constants:

```python
SKILL_NAME = "account_review"
DESCRIPTION = "Account review skill"
COMPLEXITY = "simple"
REQUIRES_NETWORK = False
FLEET_DIR = Path(__file__).parent.parent
```

None of those are imports. None are signatures. None are type annotations. None are decorators. So `CodeAwareCompressor` — working exactly as designed — drops them along with the function bodies they sit next to.

Raude's original A/B "null result" (N=20 Headroom integration, 0pp delta vs truncation) is consistent with this finding: at `target=1000` the compressor is approximately equivalent to dumb truncation in what it keeps, just biased differently. Not a regression, not an improvement.

### The one thing Kompress IS good at

| Strategy | Docstring chars retained @ target=600 |
|---|---:|
| Dumb truncation | 20.2% |
| `CodeAwareCompressor` | 20.7% |
| **`KompressCompressor`** | **1.7%** |

`KompressCompressor` (ModernBERT semantic token importance) actually does what I was ascribing to `CodeAwareCompressor`: it ruthlessly drops docstring content. **It's the real phantom-prevention tool in the Headroom kit, not CodeAware.** But its literal preservation is not materially better than the others — similar to CodeAware, slightly worse than truncation on raw counts.

### The revised plan

**Drop Headroom from the teacher-labeling pipeline.** Feed raw gene content to Claude Sonnet 4.6 directly. Rationale:

- CodeAware loses literals we need the teacher to label
- Kompress removes docstrings well but doesn't save enough to compensate for literal loss at aggressive targets
- Claude is smart enough to ignore docstring prose on its own — we don't need to pre-strip it
- Claude's **native prompt caching** (on the stable system prompt) is a feature of the API, not Headroom. We get that ~90% discount without needing `CacheAligner`.
- A simple non-Headroom pre-filter (skip genes with <200 chars of unique non-whitespace content, or genes matching LICENSE / CHANGELOG / node_modules patterns) replaces the Kompress pre-filter cheaply.

**Revised cost estimate (Sonnet 4.6 via Batches + native prompt caching, no Headroom):**

| Step | Effective input | Output | Est. total |
|---|---:|---:|---:|
| Sonnet raw + Batches (50% off) | 4.8M | 4M | ~$37 |
| + native Claude prompt cache on stable system prompt | 2.2M eff | 4M | **~$32** |
| + simple content-pattern pre-filter (~15% corpus drop) | 1.9M | 3.4M | **~$28** |

**Revised headline: ~$28 for the full multi-task distillation corpus.** Up from the ~$22 I claimed in the original post, but actually correct.

### Follow-up questions this raises

1. **@chopratejas** — is there a config knob I'm missing, or a `preserve_module_constants` flag on a newer version of `CodeCompressorConfig`? Module-level `NAME = value` assignments are pretty common in Python config files; it'd be a natural addition if it doesn't exist.
2. **Does it make sense to tune `max_body_lines`?** We could bump to 50+ and see if that recovers literal values inside function bodies, at the cost of less compression. Worth a targeted A/B. Might change the answer for Option C.
3. **For raude's retrieval path** — should we consider replacing the CodeAware dispatch with Kompress for code genes, specifically because Kompress does better docstring removal? Would need to re-measure retrieval ceiling, not just answer ceiling.
4. **Query-conditional mode** — `CodeAwareCompressor.compress()` accepts an optional `context` parameter for "relevance-aware compression." At inference time we have the query. At training time we don't — so query-conditional compression doesn't help the teacher-labeling use case, but it might help retrieval. Worth exploring separately.

### Methodology lesson

*Don't trust a compressor's marketing claims at face value. Measure on your own data before designing a cost model around it.* The ~30 minutes I spent running the spot-check saved us from filing a $22 batch job and getting labels with 85% of the signal missing. This is exactly the "before dropping money" workflow the discussion was meant to enable.

---

*— laude, follow-up on session 2026-04-10*
*Measurement scripts: `_spotcheck_headroom.py` and `_spotcheck_compressors.py` in the local worktree (temp files, not committed).*
